tatanka/tcp/server.go

```diff
@@ -0,0 +1,210 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package tcp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	clientcomms "decred.org/dcrdex/client/comms"
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/msgjson"
+	"decred.org/dcrdex/server/comms"
+	"decred.org/dcrdex/tatanka/mj"
+	"decred.org/dcrdex/tatanka/tanka"
+)
+
+// linkWrapper wraps a comms.Link to create a tanka.Sender.
+type linkWrapper struct {
+	comms.Link
+}
+
+func (l *linkWrapper) Request(msg *msgjson.Message, respHandler func(*msgjson.Message)) error {
+	const requestTimeout = time.Second * 30
+	return l.Link.Request(msg, func(_ comms.Link, respMsg *msgjson.Message) {
+		respHandler(respMsg)
+	}, requestTimeout, func() { // expire
+		errMsg := mj.MustResponse(msg.ID, nil, msgjson.NewError(mj.ErrTimeout, "request timed out"))
+		respHandler(errMsg)
+	})
+}
+
+func (l *linkWrapper) RequestRaw(msgID uint64, rawMsg []byte, respHandler func(*msgjson.Message)) error {
+	const requestTimeout = time.Second * 30
+	return l.Link.RequestRaw(msgID, rawMsg, func(_ comms.Link, respMsg *msgjson.Message) {
+		respHandler(respMsg)
+	}, requestTimeout, func() { // expire
+		errMsg := mj.MustResponse(msgID, nil, msgjson.NewError(mj.ErrTimeout, "request timed out"))
+		respHandler(errMsg)
+	})
+}
+
+func (l *linkWrapper) PeerID() (peerID tanka.PeerID) {
+	copy(peerID[:], []byte(l.CustomID()))
+	return
+}
+
+func (l *linkWrapper) SetPeerID(peerID tanka.PeerID) {
+	l.SetCustomID(string(peerID[:]))
+}
+
+type wsConnWrapper struct {
+	clientcomms.WsConn
+	cm *dex.ConnectionMaster
+	id atomic.Value
+}
+
+func newWsConnWrapper(cl clientcomms.WsConn, cm *dex.ConnectionMaster) *wsConnWrapper {
+	w := &wsConnWrapper{
+		WsConn: cl,
+		cm:     cm,
+	}
+	w.id.Store(tanka.PeerID{})
+	return w
+}
+
+func (cl *wsConnWrapper) Disconnect() {
+	cl.cm.Disconnect()
+	cl.cm.Wait()
+}
+
+func (cl *wsConnWrapper) SetPeerID(id tanka.PeerID) {
+	cl.id.Store(id)
+}
+
+func (cl *wsConnWrapper) PeerID() tanka.PeerID {
+	return cl.id.Load().(tanka.PeerID)
+}
+
+// Server handles HTTP, HTTPS, WebSockets, and Tor communications protocols.
+type Server struct {
+	t   TankaCore
+	ctx context.Context
```

Comment:
Srry to be a broken record, but ctx looks unused currently, can we leave it out of the struct?